### ISSUE TITLE
support for adding device by name

### ISSUE THREAD

**Issue by mogorman** [agittins/bermuda#229]
**Is your feature request related to a problem? Please describe.**
My headphones generate a random mac address every boot.  making them difficult to track.
**Describe the solution you'd like**
The mac address changes every time, but the name is always the same. so i would like to be able to track by name like we can track by uuid.

**Describe alternatives you've considered**
I could make an automation that handles this, but it would be gross

**Comment by agittins** [comment]
> > Describe alternatives you've considered

> I could make an automation that handles this, but it would be gross

🤣 

I agree this is a pretty solid use-case.

I'm thinking that implementing this as a regex but defaulting to a case-insensitive substring match will be a reasonable solution, probably as a multi-line text area in the config. So someone can put a full regex in if they have particular needs, or they can simply put the plain name and either way it will match.

There are a few important caveats, but I don't think they're so severe that this isn't worth doing:
- Multiple devices of the same model won't be distinguishable, whether they are cohabitants, neighbours or passing traffic. So it might give some bogus info from time to time, but this is still better than not having any info.
- Device names are not always provided in the advertisement, but via followup query-response packets. This means that a device might be present but we haven't discovered its name yet, so wouldn't be able to tag and act on it appropriately. This might lead to delays in "seeing" the device from Bermuda's perspective. But again, better than not tracking it at all.

So I am keen to implement this, and it sort of fits into how Bermuda treats iBeacons and IRK(Private BLE) devices as "meta-devices", so the bones of doing it are there, I just need to be able to implement it without breaking anything :-)

**Comment by jleinenbach** [comment]
I don't know the reason, but the "names" of my eufy SmartTrack Cards can be seen by Home Assistant's Bluetooth LE Tracker, but not by Bermuda. All these cards begin with "t87b2_", the rest is individually static for each card and not changeable by the user. They have random MACs, too.

device_tracker:
  - platform: bluetooth_le_tracker
    track_new_devices: true

**Comment by agittins** [comment]
@jleinenbach interesting - if you upload the result of a `download diagnostics` I'll see what I'm missing. If you don't want to upload the entire file you could probably trim down the diagnostic to just include the relevant `devices` entry and the bluetooth manager `scanners` entry for the device on one of the scanners (but feel free to upload the whole thing).

For the record it's not entirely related to this issue (it's relevant to the cause, but only partially to the solution, if that makes sense) but I'd still like to look into it.

**Comment by jleinenbach** [comment]
Thank you, but I guess this won't fix.
(I could get this with the debug logs. After I tried to "download the diagnostics", one core of the container was at 100% and the HA container didn't respond anymore and I didn't try again as I think the debug logs are enough.

The reason seems to be that my card's name gets visible when I "call/search" it:

```
2024-07-30 16:19:55.765 DEBUG (MainThread) [custom_components.bermuda] New Advert! change: BluetoothChange.ADVERTISEMENT, scanner: 48:27:E2:XX:XX:XX mac: CC:4F:32:1A:B6:DC name: T87B2_F550 serviceinfo: <BluetoothServiceInfoBleak name=T87B2_F550 address=CC:4F:32:1A:B6:DC rssi=-79 manufacturer_data={} service_data={} service_uuids=[] source=48:27:E2:E3:FB:38 connectable=True time=1758026.579326074 tx_power=None>
```

....

But after a while, the name seems to be replaced by its random mac address:

```
2024-07-30 16:20:53.346 DEBUG (MainThread) [custom_components.bermuda] New Advert! change: BluetoothChange.ADVERTISEMENT, scanner: E4:65:B8:XX:XX:XX mac: CC:4F:32:1A:B6:DC name: CC:4F:32:1A:B6:DC serviceinfo: <BluetoothServiceInfoBleak name=CC:4F:32:1A:B6:DC address=CC:4F:32:1A:B6:DC rssi=-94 manufacturer_data={} service_data={'0000fd44-0000-1000-8000-00805f9b34fb': b'8\x15\xcc\x15\xee\xd5\xdc\x1c\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'} service_uuids=[] source=E4:65:B8:XX:XX:XX connectable=False time=1758084.161319305 tx_power=None>
```

With this enhancement, I would need to make my cards visible, add them by name, and they will not be tracked until they become visible again by "calling" them.
It seems the names are visible as long as the cards are connectable=True

I guess this is a usual behavior and a security feature, but I was not aware of it.

**Comment by agittins** [comment]
> Thank you, but I guess this won't fix. (I could get this with the debug logs. After I tried to "download the diagnostics", one core of the container was at 100% and the HA container didn't respond anymore and I didn't try again as I think the debug logs are enough.

Ouch - how long do you think the diagnostics was running before you restarted it? I did find it was a bit slow on my system but maybe it is working away too hard after all. The cpu load is caused by teh redaction code, I expect. If there are lots of bluetooth addresses in the discovery history it might take a fair while to run. I'll look into that.
 
> The reason seems to be that my card's name gets visible when I "call/search" it:

Yes, some (many) devices don't send their name in the original advert, instead they wait until a device sends them a `scan_info` packet, then they respond with their name and some other stuff. From our end what that looks like is the device just has a mac, then at some point it gets a name.

Since yours reverts to a mac then I consider that to be a bug in Bermuda. Since the name comes from the device after the proxy requests it, only proxies with `active` scanning (not to be confused with active proxying) will tend to 'discover' the name. So proxies that are not in active mode, or those who are close enough to hear the advert but too far away to succeed at a scan-info exchange will not report a name.

What Bermuda does is try to come up with a sensible name to show for the device by checking *each* proxy, and if the info changes Bemuda will update it. This normally works ok, but clearly in some situations it does not!

> With this enhancement, I would need to make my cards visible, add them by name, and they will not be tracked until they become visible again by "calling" them. It seems the names are visible as long as the cards are connectable=True

How often does your device change its mac? Many use a 'random mac' but they might only randomise it once in their entire service life, or just once per battery replacement. I wouldn't expect many devices designed for static, at-home use to rotate their macs at all, since it would make connecting to them difficult.

**Comment by jleinenbach** [comment]
Thank you for your explanations. 

> Ouch - how long do you think the diagnostics was running before you restarted it? I did find it was a bit slow on my system but maybe it is working away too hard after all. The cpu load is caused by teh redaction code, I expect. If there are lots of bluetooth addresses in the discovery history it might take a fair while to run. I'll look into that.

Around 10 minutes. HA stopped logging and the web GUI lost the connection.

> How often does your device change its mac? Many use a 'random mac' but they might only randomise it once in their entire service life, or just once per battery replacement. I wouldn't expect many devices designed for static, at-home use to rotate their macs at all, since it would make connecting to them difficult.

I am not sure. I added them to Bermuda and the MAC address was a different one the next day.

**Comment by agittins** [comment]
> Around 10 minutes. HA stopped logging and the web GUI lost the connection.

I'm sorry about that. For it to get that badly wedged my guess is that it might have exceeded the available memory/stack. I used a recursive function to traverse the tree of diagnostic info, which means the memory usage increases with tree depth. I think I'll need to do it differently.

> the MAC address was a different one the next day.

Yes, that's annoying! Well if I can fix the issue of Bermuda *dropping* the name, then perhaps the match-by-name feature might work for you, too.

**Comment by mobs99** [comment]
Hi, just wondering if there is been any update on this enhancement, looking for this exact feature at the moment.

**Comment by agittins** [comment]
Yes this feature request is getting pretty long in the tooth now! I'll see if I can prioritise this for the next round of work.

**Comment by Nickduino** [comment]
> * Device names are not always provided in the advertisement, but via followup query-response packets. This means that a device might be present but we haven't discovered its name yet, so wouldn't be able to tag and act on it appropriately. This might lead to delays in "seeing" the device from Bermuda's perspective. But again, better than not tracking it at all.

Only the first time a random MAC is generated, right? After that, I assume you'd track the MAC address (and possibly stop after a few days because it's safe to assume it has changed)

**Comment by agittins** [comment]
> Only the first time a random MAC is generated, right? After that, I assume you'd track the MAC address (and possibly stop after a few days because it's safe to assume it has changed)

Yes, but devices typically generate a new Mac every 15 minutes or less. Peripherals like headphones etc seem pretty keen to advertise their names so this might not be a problem for certain devices, while others are more tight-lipped.

**Comment by Nickduino** [comment]
I had no idea. My Sony XM4 keep the same MAC for days and the OP says his generate a new MAC at every boot.

I guess phones are a bit more concerned with anonymity but that's solved with the companion app iBeacon settings, so...

**Comment by DCooper-nz** [comment]
I would be keen to see this implemented so I can could track my Fitbit Air as it changes Mac address every 10 - 15 minutes.
